### PR TITLE
Port over FI_* env vars to new var API.

### LINKED ISSUE #1106: Convert remaining environment vars to use new fi_var.h API
Specifically the FI_LOG_* set.

I have a patch in my local tree that does some of this work.

Issue encountered: if we use the "core" provider to register the variables all their names will be changed from FI_LOG_* to FI_CORE_LOG_*

If this is not OK then an additional internal-only API might need to be added to keep the original names.

### PR DESCRIPTION
**Pull request by pmmccorm**
All FI_ env variables are registered using the fake core provider.
A special case is made for these to omit the CORE prefix from the env var name.

fixes #1106 

Signed-off-by: pmmccorm <patrick.m.mccormick@intel.com>

### CODE REVIEW

**Comment by pmmccorm** [comment]
Current registered vars:
```
fi_info -e
# psm: Unique Job ID required by the PSM fabric
# FI_PSM_UUID

# psm: Whether to use tagged messages for large size RMA or not (default: yes)
# FI_PSM_TAGGED_RMA

# psm: Whether to use active message based messaging or not (default: no)
# FI_PSM_AM_MSG

# psm: Whether to turn on the name server or not (default: yes)
# FI_PSM_NAME_SERVER

# sockets: Drop every Nth dgram frame (debug only)
# FI_SOCKETS_DGRAM_DROP_RATE

# sockets: How many miliseconds to spin while waiting for progress
# FI_SOCKETS_PE_WAITTIME

# core: Search for providers in specific path (default: /home/pmmccorm/local/lib/libfabric)
# FI_PROVIDER_PATH

# core: Only use specified provider (default: all available)
# FI_PROVIDER

# core: Specify specific subsystem to log (default: all)
# FI_LOG_SUBSYS

# core: Specify specific provider to log (default: all)
# FI_LOG_PROV

# core: Specify logging level: warn, trace, info, debug (default: warn)
# FI_LOG_LEVEL
```

**Comment by shefty** [comment]
Please keep 1 concept per patch.

File: src/log.c [hunk 1]
```diff
@@ -62,6 +63,8 @@ static const char * const log_levels[] = {
 	[FI_LOG_MAX] = NULL
 };
 
+static struct fi_provider core_prov = FI_CORE_PROVIDER;
```
Comment (shefty) [inline src/log.c]: Have 1 core provider and just share the definition between files.  Don't declare the same variable with the intent that it have exactly the same value twice.

File: src/var.c [hunk 2]
```diff
@@ -166,10 +166,17 @@ int DEFAULT_SYMVER_PRE(fi_var_register)(const struct fi_provider *provider,
 	v->provider = provider;
 	v->var_name = strdup(var_name);
 	v->help_string = strdup(help_string);
-	if (!v->var_name || !v->help_string || 
-		asprintf(&v->env_var_name, "FI_%s_%s",
+
+	// core providers don't get core prefix
+	if (strncmp("core", provider->name, sizeof ("core")))
```
Comment (shefty) [inline src/var.c]: This should just compare the provider pointers rather than doing a string compare

**Comment by shefty** [comment]
See PR #1113 for other changes to the API.
For the core case, I think we want to pass in NULL as the provider and check on that.  This will allow us to define variables that can apply to multiple providers, as well as the core.

File: src/log.c [hunk 3]
```diff
@@ -80,18 +83,13 @@ enum {
 uint64_t log_mask;
 struct fi_filter prov_log_filter;
 
-
-static int fi_read_value(const char *env_name, const char * const names[])
+static int fi_read_value(const char *env_name)
```
Comment (shefty) [inline src/log.c]: This would be better renamed to env_val

File: src/log.c [hunk 4]
```diff
@@ -100,16 +98,41 @@ static int fi_read_value(const char *env_name, const char * const names[])
...
 {
 	struct fi_filter subsys_filter;
-	int level, i;
+	int level, i, ret;
+	char *levelstr = NULL, *provstr = NULL, *subsysstr = NULL;
+
+	ret = fi_var_register(&core_prov, "log_level",
+		"Specify logging level: warn, trace, info, debug (default: warn)");
+
+	ret += fi_var_get_str(&core_prov, "log_level", &levelstr);
```
Comment (shefty) [inline src/log.c]: Don't add return values.  You have no idea what the final result means.

File: src/log.c [hunk 5]
```diff
@@ -100,16 +98,41 @@ static int fi_read_value(const char *env_name, const char * const names[])
...
-	level = fi_read_value("FI_LOG_LEVEL", log_levels);
+	level = fi_read_value(levelstr);
 	if (level >= 0)
 		log_mask = ((1 << (level + 1)) - 1);
 
-	fi_create_filter(&prov_log_filter, "FI_LOG_PROV");
+	ret = fi_var_register(&core_prov, "log_prov",
+		"Specify specific provider to log (default: all)");
+
+	ret += fi_var_get_str(&core_prov, "log_prov", &provstr);
```
Comment (shefty) [inline src/log.c]: same here - don't add return values.
and kill the extra blank line between assigning the value and checking with the if statement

File: src/log.c [hunk 6]
```diff
@@ -100,16 +98,41 @@ static int fi_read_value(const char *env_name, const char * const names[])
...
+		fprintf(stderr, "failed to initialize logging provider from FI_LOG_PROV\n");
+
+	fi_create_filter(&prov_log_filter, provstr);
 	/* providers are selectively disabled */
 
-	fi_create_filter(&subsys_filter, "FI_LOG_SUBSYS");
+	ret = fi_var_register(&core_prov, "log_subsys",
+		"Specify specific subsystem to log (default: all)");
+
+	ret += fi_var_get_str(&core_prov, "log_subsys", &subsysstr);
```
Comment (shefty) [inline src/log.c]: You do realize that this is almost too ugly for words, right?  : P

**Comment by pmmccorm** [comment]
Updated with Sean's input. 20% less unspeakably ugly :)

Core variable registration and lookup happens with NULL as the provider (so other providers can lookup "global" env vars by passing NULL).

fi_getsettings() will sub in "core" for the provider name when encountering a NULL provider.

**Comment by shefty** [comment]
This conflicts with the changes in my PR, so I will merge manually, pending any comments against my changes.

**Comment by pmmccorm** [comment]
One thing I forgot to mention:

This won't work:
```
FI_PROVIDER=$someprov fi_info -e
```

There may be an easy fix but the real fix will boil down to: we need a var unregistering API and the providers need to call it from their cleanup methods.

**Comment by shefty** [comment]
Yes, I think we should add an unregister call.  Let me merge everything, and I'll add that before creating the rc.

**Pull request closed without merging**